spring-pulsar/src/main/java/org/springframework/pulsar/core/DefaultTopicResolver.java

```diff
@@ -70,28 +69,40 @@ public Map<Class<?>, String> getCustomTopicMappings() {
 	}
 
 	@Override
-	public Optional<String> resolveTopic(@Nullable String userSpecifiedTopic, Supplier<String> defaultTopicSupplier) {
-		return doResolveTopic(userSpecifiedTopic, null, defaultTopicSupplier);
+	public Resolved<String> resolveTopic(@Nullable String userSpecifiedTopic, Supplier<String> defaultTopicSupplier) {
+		if (StringUtils.hasText(userSpecifiedTopic)) {
+			return Resolved.of(userSpecifiedTopic);
+		}
+		String defaultTopic = defaultTopicSupplier.get();
+		if (defaultTopic == null) {
+			return Resolved.failed("Topic must be specified when no default topic is configured");
+		}
+		return Resolved.of(defaultTopic);
 	}
 
 	@Override
-	public <T> Optional<String> resolveTopic(@Nullable String userSpecifiedTopic, T message,
+	public <T> Resolved<String> resolveTopic(@Nullable String userSpecifiedTopic, @Nullable T message,
 			Supplier<String> defaultTopicSupplier) {
-		return doResolveTopic(userSpecifiedTopic, message.getClass(), defaultTopicSupplier);
+		return doResolveTopic(userSpecifiedTopic, message != null ? message.getClass() : null, defaultTopicSupplier);
 	}
 
 	@Override
-	public Optional<String> resolveTopic(@Nullable String userSpecifiedTopic, Class<?> messageType,
+	public Resolved<String> resolveTopic(@Nullable String userSpecifiedTopic, @Nullable Class<?> messageType,
 			Supplier<String> defaultTopicSupplier) {
 		return doResolveTopic(userSpecifiedTopic, messageType, defaultTopicSupplier);
 	}
 
-	private Optional<String> doResolveTopic(@Nullable String userSpecifiedTopic, @Nullable Class<?> messageType,
+	private Resolved<String> doResolveTopic(@Nullable String userSpecifiedTopic, @Nullable Class<?> messageType,
```

Comment:
Should we open visiblity to protected here like we did for SR?

Reply:
done